File: iosApp/iosApp/Features/Home/Components/RecentTransactionsSection.swift
import SwiftUI
import Shared

/// ホーム画面で表示する「最近の取引履歴」1 件分の表示モデル。
///
/// `direction` から sign（`+`/`-`）と金額色（ピンク/黒）を派生させる。Android 側
/// `RecentTransactionItem` と意味論を揃えてある。名前解決 / アーティファクト名取得は
/// 後続タスクで対応するため、`title` は呼び出し側で短縮 ID を組み込んで作成する。
struct RecentTransactionItem: Identifiable {
    let id: String
    let title: String
    let amount: Int64
    let direction: TransactionDirection
    let timestamp: String
}

extension RecentTransactionItem {
    /// shared `Transaction` を表示用 `RecentTransactionItem` に変換する。
    /// タイトル組み立ては `TransactionRow.swift` の `TransactionRowVariant` と同じロジック。
    static func fromShared(_ transaction: Shared.Transaction) -> RecentTransactionItem {
        let direction = transaction.direction
        let title: String
        if direction == TransactionDirection.mint {
            let suffix = transaction.artifactId.map { String($0.suffix(TransactionDisplay.shortIdLength)) }
            title = suffix.map { "アーティファクト \($0)" } ?? "発行"
        } else if direction == TransactionDirection.incoming {
            title = transaction.counterpartyPublicId
                .map { "@\($0) からもらいました" } ?? "入金"
        } else {
            title = transaction.counterpartyPublicId
                .map { "@\($0) に送りました" } ?? "送金"
        }
        let timestamp = TransactionDateFormatterIosKt.formatTransactionDateTimeSlashForIos(instant: transaction.occurredAt)
        return RecentTransactionItem(
            id: transaction.id,
            title: title,
            amount: transaction.amount,
            direction: direction,
            timestamp: timestamp,
        )
    }
}

/// ホーム画面の「最近の取引履歴」セクション — Figma `709:8658` 準拠。
///
/// セクションヘッダー（タイトル + もっとみる）と、白背景・角丸 20 の取引カード 3 枚を
/// 縦に並べる。
struct RecentTransactionsSection: View {
    let items: [RecentTransactionItem]
    let onMore: () -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            sectionHeader
            VStack(spacing: 2) {
                ForEach(items) { item in
                    RecentTransactionCard(item: item)
                }
            }
            .frame(maxWidth: .infinity)
        }
        .padding(.horizontal, 4)
        .padding(.vertical, 8)
        .frame(maxWidth: .infinity)
    }

    private var sectionHeader: some View {
        HStack {
            Text("最近の取引履歴")
                .font(FujuBankTypography.sectionLabel)
                .foregroundStyle(FujuBankPalette.textPrimary)
            Spacer()
            Button(action: onMore) {
                HStack(spacing: 4) {
                    Text("もっとみる")
                        .font(FujuBankTypography.linkAction)
                        .foregroundStyle(FujuBankPalette.linkBlue)
                    Image("ChevronRight")
                        .renderingMode(.template)
                        .resizable()
                        .scaledToFit()
                        .frame(width: 14, height: 14)
                        .foregroundStyle(FujuBankPalette.linkBlue)
                }
            }
            .buttonStyle(.plain)
        }
        .padding(.horizontal, 8)
    }
}

private struct RecentTransactionCard: View {
    let item: RecentTransactionItem

    var body: some View {
        // sign と金額色は direction から派生させる。Outgoing は黒/`-`、Mint/Incoming はピンク/`+`。
        let isOutgoing = item.direction == TransactionDirection.outgoing
        let sign = isOutgoing ? "-" : "+"
        let amountColor: Color = isOutgoing ? FujuBankPalette.textPrimary : FujuBankPalette.brandPink
        VStack(alignment: .leading, spacing: 4) {
            HStack(alignment: .top) {
                Text(item.title)
                    .font(FujuBankTypography.title)
                    .foregroundStyle(FujuBankPalette.textPrimary)
                Spacer()
                Text("\(sign)\(CurrencyFormatter.shared.formatAmount(amount: item.amount)) \(CurrencyFormatter.shared.UNIT)")
                    .font(FujuBankTypography.rowAmount)
                    .foregroundStyle(amountColor)
            }
            Text(item.timestamp)
                .font(FujuBankTypography.caption)
                .foregroundStyle(FujuBankPalette.textSecondary)
                .frame(maxWidth: .infinity, alignment: .leading)
        }
        .padding(16)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(FujuBankPalette.surface)
        .clipShape(RoundedRectangle(cornerRadius: 20))
        .shadow(color: FujuBankPalette.shadowTint.opacity(0.06), radius: 4, x: 0, y: 2)
    }
}

File: composeApp/src/androidMain/kotlin/studio/nxtech/fujubank/features/home/components/RecentTransactionsSection.kt
package studio.nxtech.fujubank.features.home.components

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.ColorFilter
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import studio.nxtech.fujubank.R
import studio.nxtech.fujubank.domain.model.TransactionDirection
import studio.nxtech.fujubank.format.CurrencyFormatter
import studio.nxtech.fujubank.theme.FujuBankColors
import studio.nxtech.fujubank.theme.NotoSansJP

/**
 * ホーム画面の「最近の取引履歴」セクション — Figma `709:8658` 準拠。
 *
 * セクションヘッダー（タイトル + もっとみる）と、白背景・角丸 20dp の取引カード 3 枚を縦に並べる。
 * バックエンド連携は本タスクのスコープ外のため、表示するアイテムは呼び出し側からモックを渡す。
 */
@Composable
fun RecentTransactionsSection(
    items: List<RecentTransactionItem>,
    onMore: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier = modifier
            .fillMaxWidth()
            .padding(horizontal = 4.dp, vertical = 8.dp),
        verticalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        SectionHeader(onMore = onMore)
        Column(
            modifier = Modifier.fillMaxWidth(),
            verticalArrangement = Arrangement.spacedBy(2.dp),
        ) {
            items.forEach { item ->
                RecentTransactionCard(item = item)
            }
        }
    }
}

@Composable
private fun SectionHeader(onMore: () -> Unit) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 8.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.SpaceBetween,
    ) {
        Text(
            text = "最近の取引履歴",
            style = TextStyle(
                fontFamily = NotoSansJP,
                fontSize = 12.sp,
                fontWeight = FontWeight.Bold,
                color = FujuBankColors.TextPrimary,
            ),
        )
        Row(
            modifier = Modifier.clickable(onClick = onMore),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(4.dp),
        ) {
            Text(
                text = "もっとみる",
                style = TextStyle(
                    fontFamily = NotoSansJP,
                    fontSize = 12.sp,
                    fontWeight = FontWeight.Bold,
                    color = FujuBankColors.LinkBlue,
                ),
            )
            Image(
                painter = painterResource(R.drawable.ic_chevron_right),
                contentDescription = null,
                modifier = Modifier.size(14.dp),
                colorFilter = ColorFilter.tint(FujuBankColors.LinkBlue),
            )
        }
    }
}

@Composable
private fun RecentTransactionCard(item: RecentTransactionItem) {
    // sign と金額色は direction から派生させる。Outgoing は黒/`-`、Mint/Incoming はピンク/`+`。
    val sign = if (item.direction == TransactionDirection.Outgoing) "-" else "+"
    val amountColor = if (item.direction == TransactionDirection.Outgoing) {
        FujuBankColors.TextPrimary
    } else {
        FujuBankColors.BrandPink
    }
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .shadow(
                elevation = 4.dp,
                shape = RoundedCornerShape(20.dp),
                clip = false,
            )
            .clip(RoundedCornerShape(20.dp))
            .background(FujuBankColors.Surface)
            .padding(16.dp),
        verticalArrangement = Arrangement.spacedBy(4.dp),
    ) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            verticalAlignment = Alignment.Top,
            horizontalArrangement = Arrangement.SpaceBetween,
        ) {
            Text(
                text = item.title,
                style = TextStyle(
                    fontFamily = NotoSansJP,
                    fontSize = 14.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = FujuBankColors.TextPrimary,
                ),
            )
            Text(
                text = "${sign}${CurrencyFormatter.formatAmount(item.amount)} ${CurrencyFormatter.UNIT}",
                style = TextStyle(
                    fontFamily = NotoSansJP,
                    fontSize = 15.sp,
                    fontWeight = FontWeight.Bold,
                    color = amountColor,
                ),
            )
        }
        Text(
            text = item.timestamp,
            modifier = Modifier.fillMaxWidth(),
            style = TextStyle(
                fontFamily = NotoSansJP,
                fontSize = 12.sp,
                fontWeight = FontWeight.Normal,
                color = FujuBankColors.TextSecondary,
            ),
        )
    }
}

/**
 * ホーム画面で表示する「最近の取引履歴」1 件分の表示モデル。
 *
 * `direction` から sign（`+`/`-`）と金額色（ピンク/黒）を派生させる。
 * 名前解決 / アーティファクト名取得は後続タスクで対応するため、`title` は呼び出し側で
 * 短縮 ID を組み込んで作成する（取引一覧の `TransactionRow` と同じロジック）。
 */
data class RecentTransactionItem(
    val title: String,
    val amount: Long,
    val direction: TransactionDirection,
    val timestamp: String,
)
